### ISSUE TITLE
grid of thumbnails showing step options in addStep ui

### ISSUE THREAD

**Issue by jywarren** [publiclab/image-sequencer#247]
We could generate a tiny thumbnail of the final image and run 5 or so of the most popular or useful modules, showing very low res previews of what they'd look like. Sort of like in Instagram!

(Low priority but fun, also dependent on a scale module)

**Comment by jywarren** [comment]
![screenshot_20180504-224813](https://user-images.githubusercontent.com/24359/39659015-4756e8a2-4fed-11e8-8a62-4a65c735657a.png)

**Comment by JonathanXu1** [comment]
This would be a cool task to work on for GCI! I'm happy to work on it if it will be made into one :)

**Comment by jywarren** [comment]
This could be done as an alternative tab to the current "icons" ui approach. Or maybe you could think of another way it could be implemented! Perhaps so as not to have to auto-generate all the thumbnails unless you really want to?

Creating a GCI task now!

**Comment by jywarren** [comment]
https://codein.withgoogle.com/dashboard/tasks/6720201132343296/

**Comment by JonathanXu1** [comment]
Thanks for the speedy response! I'm thinking of replacing the icons with this, with a crop overlay above the image for a crop preview. So which adjustments would you like previews for?

**Comment by jywarren** [comment]
Oh, interesting! I do think we should preserve the ability to not generate
the thumbnails depending on how intensive it is.

we should use the default settings for each step, and this may mean we need
to tweak them to be more as expected .... i.e. how blurry is default blur?
how bright is default brightness?

Would you like to start with the initial set of quick steps? Saturation
seems like a fun one.

In general, perhaps we should start with a method which can generate a
preview version of an image, given any arbitrary step? I this this could
involve just creating a new sequencer object and passing it an image...

**Comment by JonathanXu1** [comment]
I've never really worked with the image-sequencer ui before, though it is well documented. I made a new previewSequencer object in demo.js and set a custom ui for it, like so:

```javascript
  var previewSequencer = ImageSequencer();

  // UI for the preview sequencer:
  previewSequencer.setUI({
    onSetup: function(step){
      // Create new property "step.image"
      step.image = document.createElement('img');
      if($("#" + step).length() > 0){
        $("#" + step).append(step.image);
        console.log(step + " element exists");
      } else {
        console.log(step + " element does not exist");
      }
    },
    onComplete: function(step){
      // Access predefined "step.output" and user-defined "step.image"
      step.image.src = step.output;
    },
    onRemove: function(step){
      // Access user-defined "step.image"
      step.image.remove();
    }
  });
```

Am I going in the right direction? Not really sure how to call the ui for a saturation layer for example and load the image in a div. I don't know if I should do that at all 😅

**Comment by JonathanXu1** [comment]
I've never really worked with the image-sequencer ui before, though it is well documented. I made a new previewSequencer object in demo.js and set a custom ui for it, like so:

```javascript
  var previewSequencer = ImageSequencer();

  // UI for the preview sequencer:
  previewSequencer.setUI({
    onSetup: function(step){
      // Create new property "step.image"
      step.image = document.createElement('img');
      if($("#" + step).length() > 0){
        $("#" + step).append(step.image);
        console.log(step + " element exists");
      } else {
        console.log(step + " element does not exist");
      }
    },
    onComplete: function(step){
      // Access predefined "step.output" and user-defined "step.image"
      step.image.src = step.output;
    },
    onRemove: function(step){
      // Access user-defined "step.image"
      step.image.remove();
    }
  });
```

Am I going in the right direction? Not really sure how to call the ui for a saturation layer for example and load the image in a div. I don't know if I should do that at all 😅

**Comment by jywarren** [comment]
Hm, so I think rather than building a UI, you could make new images outside -- so something like this:

```js
function generatePreview(element, steps) {
  function insertImage(src) {
    var img = document.createElement('img');
    img.src = src;
    $(element).append(img); // maybe append img to the element passed in as a parameter?
  }
  var previewSequencer = ImageSequencer();
  previewSequencer.loadImage(step_image_src)
                               .addSteps('resize:{x: 100, y: 100},' + steps)
                               .run(insertImage);
}
```

What do you think? This is probably not exactly right, but a decent start...

**Comment by JonathanXu1** [comment]
Hey, thanks so much for the clarification. I've created the function as outlined, but I'm not sure how to get images from the sequencer and pass it to the insertImage function.
https://github.com/publiclab/image-sequencer/blob/6377dbe8d5de37a1722d0fedca00a6b65ec087a7/examples/demo.js#L28-L45

 It might be related, but I'm also getting errors on that line of code:
![image](https://user-images.githubusercontent.com/22998430/49312048-e0ef4a00-f4b0-11e8-81db-154d606b65a4.png)

**Comment by jywarren** [comment]
In this case, you can probably just get the last image on the page using
CSS selectors in jQuery. But there are multiple ways to get the image -
that's probably just the fastest -- $('.step:last img')[0].src -- or maybe
a more specific selector?

We should actually have a faster way to access the output image. Perhaps we
should create a set of faster access methods to simplify things. The way to
get it within sequencer is:

sequencer.images.image1.steps[0].output.src

That's pretty long. What if it were "sequencer.output()" or
"sequencer.getOutput()" and it would return an array of outputs if there is
more than one image in the sequencer?

Jonathan, if you'd like to make a new issue for this idea, we could work on
it later. For now, the jQuery approach should be a short-term solution
perfectly adequate for this use.

**Comment by JonathanXu1** [comment]
Yeah I think getting a much more formalized way to get sequencer output would be much appreciated.

Here's my error when I use **$('.step:last img')[0].src**
![image](https://user-images.githubusercontent.com/22998430/49329394-3298e280-f54c-11e8-92b3-3e1979de20c0.png)

Here's the error when I use **previewSequencer.images.image1.steps[0].output.src**
![image](https://user-images.githubusercontent.com/22998430/49329454-18133900-f54d-11e8-854c-768104d35232.png)

Logically, I think it should work. I've pushed my revisions to this branch.

**Comment by JonathanXu1** [comment]
@jywarren @publiclab/reviewers Sorry I have 6 hours left on this task..

**Comment by JonathanXu1** [comment]
What's interesting to note is that the images object does exist:

![image](https://user-images.githubusercontent.com/22998430/49416245-55dbb180-f747-11e8-8052-9d4bc001b40c.png)

However, it is printed after the error on line 47, which is technically above it. So I'm thinking this may be due to the fact that the function to load the object starts running before it has completely loaded? I'm not sure if that's the case, but I'll look into ways of waiting to run the function until after the object has completely loaded.

**Comment by jywarren** [comment]
I'm very sorry, I didn't have a moment to check in on this today and will
do my best to help out in the morning! Thanks for your patience!

**Comment by JonathanXu1** [comment]
No worries, you've got a lot on your plate! Good night.

**Comment by kevinzluo** [comment]
Hi @JonathanXu1 
Since you've claimed this task https://github.com/publiclab/plots2/issues/2866 for GCI, would you mind letting me take a shot at this one? I can't guarantee I can even get half as far as you did, this seems really interesting.

Thanks,
Kevin Luo

**Comment by jywarren** [comment]
Hmm, maybe @kevinzluo could help move this along... in any case whoever takes it on please work collaboratively! https://github.com/publiclab/image-sequencer/pull/504

**Comment by jywarren** [comment]
I think this has to do with asynchronous events, and we may have to execute the steps that are throwing an error in a callback, which Image Sequencer allows for on some of it's methods (see docs). So, the callback is a passed function which would run once the sequencer is done... 

Are you working off the latest code in https://github.com/publiclab/image-sequencer/pull/504 ?

**Comment by kevinzluo** [comment]
Hi!

I haven't had a chance to really look at the code until this morning, and my initial idea is a bit different from what @JonathanXu1 is suggesting. Should I open up a new PR?

Also, could you please approve my task here (https://codein.withgoogle.com/dashboard/task-instances/5423277104168960/) so that I may go ahead and start working on this one?

Thanks!

**Comment by JonathanXu1** [comment]
Yep, I'll push again to make sure.

**Comment by okonek** [comment]
@jywarren Hi I've added that feature in this PR: https://github.com/publiclab/image-sequencer/pull/535 If I misunderstood your explanation or there is something to change in my solution, please let me know.
Thanks.

**Comment by JonathanXu1** [comment]
I've made pull request https://github.com/jywarren/image-sequencer/pull/11 that applies my changes on top of @okonek 's. @jywarren can merge both our forks, or @okonek can apply my changes to his fork.
